cmd/stack_create_test.go

```diff
@@ -23,291 +22,129 @@ import (
-
-	testStackName := "testing_stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stack"
-
-	args := []string{"stack", "create", testStackName}
-	_, err := cmdtest.RunAppsody(sandbox, args...)
-
-	if err == nil {
-		t.Fatal(err)
-	}
-
-	exists, err := cmdtest.Exists(testStackName)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if exists {
-		// It SHOULDN'T exist, but it might
-		err = os.RemoveAll(testStackName)
-		if err != nil {
-			t.Fatal(err)
-		}
-		t.Fatal(err)
+func TestStackCreateInvalidCases(t *testing.T) {
+	var stackCreateInvalidTests = []struct {
+		testName     string
+		stackName    string
+		args         []string // input
+		expectedLogs string   // logs that are expected in the output
+	}{
+		{"No args", "testing-stack-create-invalid", []string{"--copy", "incubator/nodej"}, "Stack name must be in the format <repo>/<stack>"},
+		{"Existing default repo", "testing-stack-create-invalid", []string{"--copy", "nodejs"}, "Stack name must be in the format <repo>/<stack>"},
+		{"Non-existing repo", "testing-stack-create-invalid", []string{"--copy", "experimental/nodejs"}, "Stack name must be in the format <repo>/<stack>"},
+		{"Badly formatted repo config", "testing-stack-create-invalid", []string{"--copy", "exp/java-microprofile"}, "Stack name must be in the format <repo>/<stack>"},
+		{"Invalid stack name underscores", "testing_stack_invalid_name", nil, "The name must start with a lowercase letter, contain only lowercase letters, numbers, or dashes, and cannot end in a dash."},
+		{"Invalid stack name length", "testing_stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stacktesting-stack", nil, "The name must be 68 characters or less"},
+		{"Invalid stack name missing", "", nil, "Invalid project-name. The name cannot be an empty string"},
+	}
+	for _, testData := range stackCreateInvalidTests {
+		// need to set testData to a new variable scoped under the for loop
+		// otherwise tests run in parallel may get the wrong testData
+		// because the for loop reassigns it before the func runs
+		tt := testData
+		// call t.Run so that we can name and report on individual tests
+		t.Run(tt.testName, func(t *testing.T) {
+			sandbox, cleanup := cmdtest.TestSetupWithSandbox(t, true)
+			defer cleanup()
+			args := append([]string{"stack", "create"}, tt.stackName)
+			if tt.args != nil {
+				args = append(args, tt.args...)
+			}
+			output, err := cmdtest.RunAppsody(sandbox, args...)
+			if err == nil {
+				t.Fatalf("Expected non-zero exit code: %v", tt.expectedLogs)
+			}
+			if !strings.Contains(output, tt.expectedLogs) {
+				t.Errorf("Did not find expected error '%s' in output", tt.expectedLogs)
+			}
+			exists, err := cmdtest.Exists(tt.stackName)
+			if err != nil {
+				t.Fatal("Error attempting to check stack exists: ", err)
+			}
+			if exists {
+				// It SHOULDN'T exist, but it might
+				err = os.RemoveAll(tt.stackName)
```

Comment:
Do we even have to remove? This is in sandbox

Reply:
The exists check should still be there

Reply:
It left a bunch of files in the testdata folder, which then failed the tests the second time running. I can look into why another time?